### PR TITLE
Refactor upload-coverage action

### PR DESCRIPTION
**Pull request by jku**
#### Summary

Refactor upload-coverage to enable upgrading the upload-artifact and download-artifact actions (as the new upload-artifact does not allow uploading with the same name several times).
* Use unique artifact names, then download all artifacts before combining them for coverage
* Upgrade the actions

#### Release Note

N/A

### CODE REVIEW

**Comment by jku** [comment]
> jku force-pushed the refactor-upload-coverage branch 10 times, ...

That took a few tries :joy: Looks fine now I think.

File: .github/actions/upload-coverage/action.yml [hunk 1]
```diff
@@ -20,9 +20,9 @@ runs:
         fi
       id: coverage-uuid
       shell: bash
-    - uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32 # v3.1.3
+    - uses: actions/upload-artifact@c7d193f32edcb7bfad88892161225aeda64e9392 # v4.0.0
       with:
-        name: coverage-data
```
Comment (jku) [inline .github/actions/upload-coverage/action.yml:25]: The artifact name could be more useful but there is a FIXME about that 10 lines above already

File: .github/actions/upload-coverage/action.yml [hunk 2]
```diff
@@ -20,9 +20,9 @@ runs:
...
+        name: coverage-data-${{ steps.coverage-uuid.outputs.COVERAGE_UUID }}
```
Comment (jku) [inline .github/actions/upload-coverage/action.yml:25]: The artifact name could be more useful but there is a FIXME about that 10 lines above already

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -91,14 +91,15 @@ jobs:
...
         with:
-          name: coverage-data
+          path: all-artifacts/
 
       - name: combine coverage data
         id: combinecoverage
         run: |
           set +e
+          cp all-artifacts/coverage-data-*/.coverage.* .
           python -m coverage combine
```
Comment (woodruffw) [inline .github/workflows/ci.yml:103]: This is a tiny thing, but I _think_ you can do this: 

```suggestion
          set +e
          python -m coverage combine all-coverage/
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -91,14 +91,15 @@ jobs:
       - run: pip install coverage[toml]
 
       - name: download coverage data
-        uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
+        uses: actions/download-artifact@7a1cd3216ca9260cd8022db641d960b1db4d1be4 # v4.0.0
         with:
-          name: coverage-data
+          path: all-artifacts/
```
Comment (woodruffw) [inline .github/workflows/ci.yml:96]: Nitpick, descriptive:

```suggestion
          path: all-coverage/
```

File: .github/workflows/ci.yml (hunk 3 above)
Comment (jku) [reply .github/workflows/ci.yml]: thanks, that's better. The argument needs to be `all-artifacts/coverage-data-*` though (download-artifacts creates extra directories when used like this).

File: .github/workflows/ci.yml (hunk 4 above)
Comment (jku) [reply .github/workflows/ci.yml:96]: I'm going to leave this unless you insist as the contents really are "all artifacts": if someone in the future adds another artifact upload earlier in the workflow, the unrelated artifact will show up in this directory. So any code that processes "all-artifacts" should be aware of that.

Comment (woodruffw) [reply .github/workflows/ci.yml:96]: Sure, works for me.

**Review by woodruffw — APPROVED** [review]

**Pull request merged**